pkg/controller/autoops/controller.go

```diff
@@ -0,0 +1,304 @@
+func reconcileRequestForAllAutoOpsPolicies(clnt k8s.Client) handler.TypedEventHandler[client.Object, reconcile.Request] {
+	return handler.TypedEnqueueRequestsFromMapFunc(func(ctx context.Context, es client.Object) []reconcile.Request {
+		var autoOpsAgentPolicyList autoopsv1alpha1.AutoOpsAgentPolicyList
+		err := clnt.List(context.Background(), &autoOpsAgentPolicyList)
+		if err != nil {
+			ulog.Log.Error(err, "Fail to list AutoOpsAgentPolicyList while watching Elasticsearch")
+			return nil
+		}
+		requests := make([]reconcile.Request, 0)
+		for _, autoOpsAgentPolicy := range autoOpsAgentPolicyList.Items {
+			requests = append(requests, reconcile.Request{NamespacedName: k8s.ExtractNamespacedName(&autoOpsAgentPolicy)})
+		}
+		return requests
+	})
+}
+
+var _ reconcile.Reconciler = (*AgentPolicyReconciler)(nil)
+
+// AgentPolicyReconciler reconciles an AutoOpsAgentPolicy object
+type AgentPolicyReconciler struct {
+	k8s.Client
+	accessReviewer   rbac.AccessReviewer
+	recorder         record.EventRecorder
+	licenseChecker   license.Checker
+	params           operator.Parameters
+	dynamicWatches   watches.DynamicWatches
+	esClientProvider commonesclient.Provider
+	// iteration is the number of times this controller has run its Reconcile method
+	iteration uint64
+}
+
+// Reconcile reconciles the AutoOpsAgentPolicy resource ensuring that any resources are created/updated/deleted as needed.
+func (r *AgentPolicyReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	ctx = common.NewReconciliationContext(ctx, &r.iteration, r.params.Tracer, controllerName, "policy_name", request)
+	defer common.LogReconciliationRun(ulog.FromContext(ctx))()
+	defer tracing.EndContextTransaction(ctx)
+
+	// retrieve the AutoOpsAgentPolicy resource
+	var policy autoopsv1alpha1.AutoOpsAgentPolicy
+	err := r.Client.Get(ctx, request.NamespacedName, &policy)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, r.onDelete(ctx,
+				types.NamespacedName{
+					Namespace: request.Namespace,
+					Name:      request.Name,
+				})
+		}
+		return reconcile.Result{}, tracing.CaptureError(ctx, err)
+	}
+
+	if common.IsUnmanaged(ctx, &policy) {
+		ulog.FromContext(ctx).Info("Object is currently not managed by this controller. Skipping reconciliation")
+		return reconcile.Result{}, nil
+	}
+
+	state := newState(policy)
+	if policy.IsMarkedForDeletion() {
+		return reconcile.Result{}, r.onDelete(ctx, k8s.ExtractNamespacedName(&policy))
+	}
+
+	results := r.doReconcile(ctx, policy, state)
+	updatePhaseFromResults(results, state)
+
+	result, err := r.updateStatusFromState(ctx, state)
+	if err != nil {
+		results.WithError(err)
+	} else if result.RequeueAfter > 0 {
+		return results.WithRequeue().Aggregate()
+	}
+
+	// requeue if not ready (similar to stackconfigpolicy controller)
+	// Check phase after status update to ensure status is persisted
+	if state.status.Phase != autoopsv1alpha1.ReadyPhase {
```

Comment:
This leads to endless re-reconciliations if we are in the NoResourcesPhase. This is unnecessary as we are watching Elasticsearch cluster anyway and will be notified through that channel if new resources are crreated.